### PR TITLE
Fix hiding issue when view switches container + Composer View Layout Polish

### PR DESCRIPTION
**Pull request by nuno-vieira**
# In this PR

- Fixes issue when a view that was already hidden would switch container, and remain with the size 0 and alpha = 0. This would cause a problem because the view would always have the size 0 and it wouldn't be possible to hide it again, because it already had alpha = 0.
- Fixes an issue on composer view that was causing the composer to layout twice
- Removes the state icon and polishes the composer layout according to Figma

With this PR, now the iMessage clone works as expected ✅

### CODE REVIEW

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerView.swift [hunk 1]
```diff
@@ -11,7 +11,7 @@ open class _ChatMessageComposerView<ExtraData: ExtraDataTypes>: _View, ThemeProv
     public private(set) lazy var container = ContainerStackView()
         .withoutAutoresizingMaskConstraints
 
-    public private(set) lazy var topContainer = ContainerStackView()
+    public private(set) lazy var topContainer = UIView()
         .withoutAutoresizingMaskConstraints
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerView.swift:15]: Why this change?

File: Sources/StreamChatUI/CommonViews/ChatInputView/ChatMessageInputView.swift [hunk 2]
```diff
@@ -36,10 +36,9 @@ open class _ChatMessageInputView<ExtraData: ExtraDataTypes>: _View, ComponentsPr
     }
     
     override open func setUpLayout() {
-        embed(container)
+        embed(container, insets: .init(top: 1, leading: 4, bottom: 1, trailing: 0))
```
Comment (VojtaStavik) [inline Sources/StreamChatUI/CommonViews/ChatInputView/ChatMessageInputView.swift:39]: The preferred way of doing this is to pin the view to the layout margins and change the margins. It makes it easier to change the layout just by changing the margins constants rather than adding new constraints.

In fact, the `inset` parameter in `embed` should be probably removed because it doesn't have a good use case.

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift [hunk 3]
```diff
@@ -128,9 +128,6 @@ open class _ChatMessageComposerVC<ExtraData: ExtraDataTypes>: _ViewController,
             dismissSuggestionsViewController()
         case let .quote(messageToQuote):
             composerView.titleLabel.text = L10n.Composer.Title.reply
-            let image = appearance.images.messageComposerReplyButton
```
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift:125]: Ah yes that makes sense 👍 Will remove it

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerView.swift (hunk 1 above)
Comment (nuno-vieira) [reply Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerView.swift:15]: image

Since there's no more left button on the top, it just easier to layout the views without the container. Or I would need to add an empty view to make the same layout. WDYT? Or is it the aim to always use the ContainerStackView? always?

File: Sources/StreamChatUI/CommonViews/ChatInputView/ChatMessageInputView.swift (hunk 2 above)
Comment (nuno-vieira) [reply Sources/StreamChatUI/CommonViews/ChatInputView/ChatMessageInputView.swift]: ✅

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift:125]: ✅

File: Sources/StreamChatUI/CommonViews/ChatInputView/ChatMessageInputView.swift (hunk 2 above)
Comment (nuno-vieira) [reply Sources/StreamChatUI/CommonViews/ChatInputView/ChatMessageInputView.swift]: I did not remove that functionality from the `embed` function btw, because it is being used in a lot of places, and it breaks the UI in multiple places. So I think it should be a ticket on its own. It would be nice to have the same API, but change the `embed` function to add the constraints to the layout margins, and the inset should be applied to the superview.layoutMargins instead

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerView.swift [hunk 4]
```diff
@@ -173,8 +169,16 @@ open class _ChatMessageComposerView<ExtraData: ExtraDataTypes>: _View, ThemeProv
         centerLeftContainer.addArrangedSubview(attachmentButton)
         centerLeftContainer.addArrangedSubview(commandsButton)
         centerLeftContainer.addArrangedSubview(shrinkInputButton)
+
+        dismissButton.widthAnchor.pin(equalToConstant: 24).isActive = true
+        dismissButton.heightAnchor.pin(equalToConstant: 24).isActive = true
```
Comment (DominikBucher12) [inline Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerView.swift:174]: nit: `dismissButton.heightAnchor.pin(equalTo: dismissButton.heightAnchor).isActive = true`

File: Sources/StreamChatUI/CommonViews/ChatInputView/ChatMessageInputView.swift [hunk 5]
```diff
@@ -36,10 +36,19 @@ open class _ChatMessageInputView<ExtraData: ExtraDataTypes>: _View, ComponentsPr
     }
     
     override open func setUpLayout() {
-        embed(container)
+        layoutMargins.left = 8
+        layoutMargins.top = 1
+        layoutMargins.right = 8
+        layoutMargins.bottom = 1
```
Comment (DominikBucher12) [inline Sources/StreamChatUI/CommonViews/ChatInputView/ChatMessageInputView.swift:42]: Maybe add comment why this is needed? :)

 I feel like from documentation of `layoutMargins`, you might be wanting to use `directionalLayoutMargins` instead... Also it says that they have the default value of 8 by default, maybe wanna give it a try? 

https://developer.apple.com/documentation/uikit/uiview/1622566-layoutmargins

**Review by DominikBucher12 — APPROVED** [review]
I have some nits mainly regarding the layoutMargins...  Otherwise this looks good to me :) Thanks for taking care of the Composer <3

File: Sources/StreamChatUI/CommonViews/ContainerStackView.swift [hunk 6]
```diff
@@ -203,6 +203,14 @@ public class ContainerStackView: UIView {
         assert(subviews.contains(subview))
         subview.removeFromSuperview()
 
+        // Reset hiding state from the removed view. This important especially if the view
+        // is added to another container, since the view will have the size to 0 and the other
+        // container won't be able to hide the view because it thinks it is already hidden.
+        subview.alpha = 1.0
+        hideConstraintsByView[subview]?.isActive = false
```
[the remark this answers is not in the document]
Comment (VojtaStavik) [reply Sources/StreamChatUI/CommonViews/ContainerStackView.swift:210]: @DominikBucher12 I don't think the constraint lifetime is defined by holding a reference to it 🤔 The `isActive` is what matters.

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift [hunk 7]
```diff
@@ -97,6 +97,12 @@ open class _ChatMessageComposerVC<ExtraData: ExtraDataTypes>: _ViewController,
         setupInputView()
     }
 
+    override open func setUpLayout() {
+        super.setUpLayout()
+
+        view.embed(composerView)
```
[the remark this answers is not in the document]
Comment (VojtaStavik) [reply Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift:103]: I don' think this established this as a rule but I personally don't like the `embed` helper and prefer using:
```
addSubview(view)
view.pin(to: self.layoutMarginsGuide)
```
since it's more explicit.

**Review by VojtaStavik — APPROVED** [review]
LGTM 👍 @DominikBucher12 has some good comments there

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerView.swift (hunk 4 above)
Comment (nuno-vieira) [reply Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerView.swift:174]: I don't like that much because you have to type more code than just doing what I did, IMO it looks cleaner than the code above, and the benefits of that are really not worth it

File: Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChatUI/ChatMessageList/MessageComposer/ChatMessageComposerVC.swift:103]: Ah yes that is way better

File: Sources/StreamChatUI/CommonViews/ContainerStackView.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChatUI/CommonViews/ContainerStackView.swift:210]: Yes exactly, the `isActive` is what is important here, the rest is just cleaning some memory
